### PR TITLE
Disabled CTE generation when snippets are used in non-SELECT queries

### LINKED ISSUE #651: error when using create table
when using `--save` and then executing a CREATE table statement, I get an error.

more context: [create-table-bug.ipynb.zip](https://github.com/ploomber/jupysql/files/11841951/create-table-bug.ipynb.zip)

### LINKED ISSUE #652: error when using `--save` and writing a CTE
(similar to https://github.com/ploomber/jupysql/issues/651, let's work on that one first)

more context: [create-table-bug-2.ipynb.zip](https://github.com/ploomber/jupysql/files/11841977/create-table-bug-2.ipynb.zip)

### PR DESCRIPTION
**Pull request by nooodle-soup**
## Describe your changes
Snippets now when used in anything but a SELECT statement gives a warning and disables CTE generation.

## Issue number

Closes #651 
Closes #652 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing)
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--878.org.readthedocs.build/en/878/

### CODE REVIEW

File: src/sql/magic.py [hunk 1]
```diff
@@ -406,22 +408,58 @@ def interactive_execute_wrapper(**kwargs):
...
+                sqlparse.parse(command.sql)[0].get_type()
+                if sqlparse.parse(command.sql)
+                else None
+            )
             if args.with_:
                 raise exceptions.UsageError(
                     "Cannot use --with with CTEs, remove --with and re-run the cell"
                 )
 
+            """
```
Comment (edublancas) [inline src/sql/magic.py:453]: This code is wrapped in a string (I guess you want to delete it?), please double check your PRs before requesting a review.

File: src/sql/magic.py [hunk 2]
```diff
@@ -406,22 +408,58 @@ def interactive_execute_wrapper(**kwargs):
         # only expand CTE if this is not a CTE itself
         if not is_cte:
             if args.with_:
+                query_type = (
+                    sqlparse.parse(command.sql)[0].get_type()
+                    if sqlparse.parse(command.sql)
+                    else None
+                )
```
Comment (edublancas) [inline src/sql/magic.py:415]: this same block (or similar one) is repeated in many places, can't we create a function? Also, I don't see any unit tests

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/sql/magic.py (hunk 1 above)
Comment (nooodle-soup) [reply src/sql/magic.py:453]: It is logic related to snippet expansion that I was not sure if I should implement yet. I'll remove it.

File: src/sql/magic.py (hunk 2 above)
Comment (nooodle-soup) [reply src/sql/magic.py:415]: On it.

**Comment by nooodle-soup** [comment]
The checks are failing for some reason I cannot understand. They're files that I haven't touched and the pytest tests run perfectly on my local machine. I am putting it up for review for some inputs @edublancas

**Comment by edublancas** [comment]
based on the logs, the image tests are failing (jupysql has some plotting capabilities). you can download the image artifacts from [here](https://github.com/ploomber/jupysql/actions/runs/6194732816?pr=878) (scroll down and you'll see an artifacts section). I downloaded them and noticed that the difference between the expected and the produced image is neglible.

my guess is that the issue is the latest matplotlib release: https://github.com/matplotlib/matplotlib/releases/tag/v3.7.3

to test this, you can try pinning the [matplotlib](https://github.com/ploomber/jupysql/blob/202f0431b313f2a8f0d5ab57ab2cc673a2ece502/setup.py#L47) dependency for a previous release. if that fixes it, please open a new issue so we fix the references images

File: src/sql/magic.py [hunk 3]
```diff
@@ -639,6 +669,79 @@ def _persist_dataframe(
...
+            self.remove_leading_parentheses(command.sql_original)
+            if original
+            else self.remove_leading_parentheses(command.sql)
+        )
+        query_type = (
+            sqlparse.parse(to_parse)[0].get_type() if sqlparse.parse(to_parse) else None
+        )
+        return query_type
+
+    def remove_leading_parentheses(self, sql_command):
```
Comment (edublancas) [inline src/sql/magic.py:686]: this issue is supposed to close

https://github.com/ploomber/jupysql/issues/651
and
https://github.com/ploomber/jupysql/issues/652

is this parenthesis removing logic relevant to those? I'm a bit confused

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/sql/magic.py (hunk 3 above)
Comment (nooodle-soup) [reply src/sql/magic.py:686]: This is only needed when someone might use a CTE in duckdb in the following manner:

```
(WITH my_penguins as (
    SELECT * from penguins.csv
) SELECT * from my_penguins)
```

Line 406-408 where we check for is_cte would fail in the above case if there is no removing the parentheses.

**Comment by nooodle-soup** [comment]
I set matplotlib to verison 3.7.2 and the tests passed. I'll open an issue regarding that.

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 ## 0.10.2dev
 
+* [Fix] Fixed snippet behaviour with queries (#651, #652)
```
Comment (edublancas) [inline CHANGELOG.md:5]: changelog entries should be more detailed.

example: `Disabled CTE generation when detecting a %%sql cell with an existing CTE `

File: src/sql/magic.py [hunk 5]
```diff
@@ -410,18 +414,43 @@ def interactive_execute_wrapper(**kwargs):
...
                 if with_:
-                    command.set_sql_with(with_)
-                    display.message(
-                        f"Generating CTE with stored snippets: {pretty_print(with_)}"
-                    )
+                    query_type = self.get_query_type(command, original=True)
+
+                    if query_type != "SELECT":
+                        display.message_warning(
+                            f"Cannot use snippets with {query_type}"
```
Comment (edublancas) [inline src/sql/magic.py:421]: the message should be clearer, something like:

```
Your query is using the identifiers 'a', 'b', 'c', which are stored snippets, but snippets only work with SELECT statements, hence, CTE generation is disabled
```

File: src/sql/magic.py [hunk 6]
```diff
@@ -410,18 +414,43 @@ def interactive_execute_wrapper(**kwargs):
...
                     with_ = None
         else:
+            query_type = self.get_query_type(command, original=True)
             if args.with_:
                 raise exceptions.UsageError(
                     "Cannot use --with with CTEs, remove --with and re-run the cell"
                 )
 
+            if query_type != "SELECT":
+                display.message_warning("Cannot use snippets in CTEs")
```
Comment (edublancas) [inline src/sql/magic.py:439]: same here, message should be clearer

File: src/sql/magic.py [hunk 7]
```diff
@@ -410,18 +414,43 @@ def interactive_execute_wrapper(**kwargs):
...
+
+            parentheses_content = re.findall(
+                r"\((.*)\)", command.sql_original.replace("\n", " ")
+            )
+            for query in parentheses_content:
+                for dependency in dependencies:
+                    if dependency in query:
+                        display.message_warning(
+                            f"Cannot use snippet {dependency} in CTE"
+                        )
```
Comment (edublancas) [inline src/sql/magic.py:453]: can you add some comments here? I'm not sure what this chunk of code is for

File: src/sql/magic.py (hunk 3 above)
Comment (edublancas) [reply src/sql/magic.py:686]: We discussed this parenthesis on a different issue, but I forgot the conclusion (I found this https://github.com/ploomber/jupysql/issues/879 but I think we discussed this somewhere else). In general, we try not to add new requirements to existing issues because it makes it harder to review the code.

also, I feel like we're mixing two separate issues here: CTE generation and SQL queries wrapped in parenthesis. these are two independent problems that we should tackle independently. so please remove this parenthesis handling logic and open an issue so we can discuss what we do about it

File: src/tests/test_magic.py [hunk 8]
```diff
@@ -1925,6 +1925,7 @@ def test_whitespaces_linebreaks_near_first_token(ip, query):
 
     ip.user_global_ns["query"] = query
     out = ip.run_cell("%sql {{query}}").result
+    print(out)
```
Comment (edublancas) [inline src/tests/test_magic.py:1928]: please check your PRs before requesting a review, this print statement looks like it was a debugging thing

File: src/tests/test_magic.py [hunk 9]
```diff
@@ -1987,3 +1988,29 @@ def test_accessing_previously_nonexisting_file(ip_empty, tmp_empty, capsys):
...
+        """%%sql
+INSERT INTO languages VALUES ('Python', 1), ('Java', 0), ('OCaml', 2)"""
+    )
+
+    ip_empty.run_cell(
+        """%%sql --save language_lt1
+select * from languages where rating < 1"""
+    )
+
+    with pytest.raises(UsageError) as _:
```
Comment (edublancas) [inline src/tests/test_magic.py:2006]: looks like this is testing the create table case but you're not testing the case when the user has a CTE. please write more tests to check for any other relevant scenarios.

File: src/sql/magic.py [hunk 10]
```diff
@@ -639,6 +669,79 @@ def _persist_dataframe(
             table_name=table_name, data_frame=frame, if_exists=if_exists, index=index
         )
 
+    def get_query_type(self, command: SQLCommand, original: bool):
```
Comment (edublancas) [inline src/sql/magic.py:672]: there's no need for this function to be an instance method, it can be a separate function. so make it an independent function and add some unit tests

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/tests/test_magic.py (hunk 8 above)
Comment (nooodle-soup) [reply src/tests/test_magic.py:1928]: Sorry about that! 😞

File: src/sql/magic.py (hunk 3 above)
Comment (nooodle-soup) [reply src/sql/magic.py]: I've removed it. Opening a new issue regarding the same.

File: src/sql/magic.py (hunk 6 above)
Comment (nooodle-soup) [reply src/sql/magic.py]: I've added better warning messages

File: src/sql/magic.py (hunk 5 above)
Comment (nooodle-soup) [reply src/sql/magic.py]: I've added better warning messages

File: src/tests/test_magic.py (hunk 9 above)
Comment (nooodle-soup) [reply src/tests/test_magic.py:2015]: Added more tests

File: src/sql/magic.py (hunk 7 above)
Comment (nooodle-soup) [reply src/sql/magic.py]: I've added comments

File: src/sql/magic.py (hunk 10 above)
Comment (nooodle-soup) [reply src/sql/magic.py]: Done

**Review by edublancas — CHANGES_REQUESTED** [review]
when addressing PR comments, don't forget to add links to the relevant changes, so it's easier for the reviewer to see the relevant pieces: https://ploomber-contributing.readthedocs.io/en/latest/contributing/responding-pr-review.html

**Comment by nooodle-soup** [comment]
I've added the links to the requested code changes.

File: src/sql/magic.py [hunk 11]
```diff
@@ -401,6 +406,9 @@ def interactive_execute_wrapper(**kwargs):
                 "is automatically set as the connection alias"
             )
 
+        # TODO: DuckDB allows queries to be wrapped in parentheses
+        # which needs to be accounted for here. If there is a (WITH ....)
+        # query, then there is an issue of is_CTE not being true.
```
Comment (edublancas) [inline src/sql/magic.py:411]: remove this. if we already have an open issue, there's no need for this note

File: src/sql/magic.py [hunk 12]
```diff
@@ -410,18 +418,57 @@ def interactive_execute_wrapper(**kwargs):
...
+                    else:
+                        command.set_sql_with(with_)
+                        display.message(
+                            f"Generating CTE with stored snippets: \
+{pretty_print(with_)}"
+                        )
                 else:
                     with_ = None
         else:
+            query_type = get_query_type(command.sql_original)
```
Comment (edublancas) [inline src/sql/magic.py:437]: you're calling get_query_type two times in two `if` branches, maybe it'd be  better to just call it at the top?

Comment (edublancas) [reply src/sql/magic.py:437]: actually, this runs when  `is_cte` is True, hence, you're guaranteed to get to this line if you're inside a CTE. do you still need to run `get_query_type`?

File: src/sql/magic.py [hunk 13]
```diff
@@ -410,18 +418,57 @@ def interactive_execute_wrapper(**kwargs):
...
+            # This is for generating more relevant warnings.
+            parentheses_content = re.findall(
+                r"\((.*)\)", command.sql_original.replace("\n", " ")
+            )
+
+            dependency_in_CTE = []
+            for query in parentheses_content:
+                for dependency in dependencies:
+                    if dependency in query:
+                        dependency_in_CTE.append(dependency)
```
Comment (edublancas) [inline src/sql/magic.py:460]: please move this to a separate function so we can unit test it.

also, this doesn't seem like the best way to parse this. check how we're inferring dependencies e.g., detect `some_table` in `SELECT * FROM some_table` using sqlglot. that sounds like a better way to parse this

File: src/sql/magic.py [hunk 14]
```diff
@@ -410,18 +418,57 @@ def interactive_execute_wrapper(**kwargs):
...
+
+            if dependency_in_CTE:
+                if query_type != "SELECT":
+                    display_message = IF_NOT_SELECT_MESSAGE
+                else:
+                    display_message = IF_SELECT_MESSAGE
```
Comment (edublancas) [inline src/sql/magic.py:466]: isn't this branch only running when `is_cte` is true? or maybe I'm missing something. under which cases you fall under one `if` condition or the other?

File: src/tests/test_magic.py [hunk 15]
```diff
@@ -1987,3 +1987,100 @@ def test_accessing_previously_nonexisting_file(ip_empty, tmp_empty, capsys):
     ip_empty.run_cell("%sql SELECT * FROM 'data.csv' LIMIT 3")
     out, _ = capsys.readouterr()
     assert expected in out
+
+
+def test_warn_when_using_snippets_in_non_SELECT_command(ip_empty, capsys):
```
Comment (edublancas) [inline src/tests/test_magic.py:1992]: looks like you're testing many things at once, use pytest's `parametrize` instead

File: src/tests/test_magic.py [hunk 16]
```diff
@@ -1987,3 +1987,100 @@ def test_accessing_previously_nonexisting_file(ip_empty, tmp_empty, capsys):
...
+    )
+    assert (
+        """Your query is using the following snippets: language_lt3. \
+The query is not a SELECT type query and as snippets only work \
+with SELECT queries, CTE generation is disabled"""
+        in out
+    )
+
+
+def test_get_query_type():
```
Comment (edublancas) [inline src/tests/test_magic.py:2066]: same here, use parametrize

File: src/tests/test_magic.py [hunk 17]
```diff
@@ -1987,3 +1987,100 @@ def test_accessing_previously_nonexisting_file(ip_empty, tmp_empty, capsys):
...
+    assert (
```
Comment (edublancas) [inline src/tests/test_magic.py:2067]: also, add one test when calling this with None or invalid SQL, in such case we should return None

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/sql/magic.py (hunk 14 above)
Comment (nooodle-soup) [reply src/sql/magic.py:466]: One case is when the CTE is used in a CREATE query like follows:

```
%%sql
WITH my_penguins AS (
    SELECT * FROM penguins.csv
) CREATE TABLE my_penguins_2 AS (
    SELECT * FROM my_penguins
)
```

The query type for this will evaluate to `CREATE`.

A regular query like the following:

```
%%sql
my_penguins AS (
    SELECT * FROM penguins.csv
) SELECT * FROM my_penguins
```

will evaluate to query type of `SELECT`.

File: src/sql/magic.py (hunk 11 above)
Comment (nooodle-soup) [reply src/sql/magic.py:411]: Will do!

File: src/sql/magic.py (hunk 12 above)
Comment (nooodle-soup) [reply src/sql/magic.py:437]: We do need to check the query type even in the case where is_cte is true as the following will evaluate to a `CREATE` query type and not `SELECT`.

```
%%sql
WITH my_penguins AS (
    SELECT * FROM penguins.csv
) CREATE TABLE my_penguins_2 AS (
    SELECT * FROM my_penguins
)
```

File: src/sql/magic.py (hunk 13 above)
Comment (nooodle-soup) [reply src/sql/magic.py:460]: I could use that method, but it would become difficult to find out which tables in the query are important to us here.

For example, if we have the following:

```
WITH outer_table AS (
    SELECT * FROM some_snippet
) SELECT * FROM outer_table
```

There are two tables here, `some_snippet` and `outer_table`.
We only want to know if `some_snippet` is a snippet or not. We do not care if `outer_table` is a snippet because we are binding it to the value of `SELECT * FROM some_snippet` in our query, so it has a value in the local scope regardless of any existing snippet in the global scope named `outer_table`. This makes it irrelevant to the error message, which is why I have checked for dependencies within each parentheses only.

File: src/sql/magic.py (hunk 11 above)
Comment (nooodle-soup) [reply src/sql/magic.py:411]: Removed the comments

File: src/tests/test_magic.py (hunk 15 above)
Comment (nooodle-soup) [reply src/tests/test_magic.py]: [Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R1992-R2036)

File: src/tests/test_magic.py (hunk 16 above)
Comment (nooodle-soup) [reply src/tests/test_magic.py]: [Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2058-R2080)

File: src/sql/magic.py (hunk 13 above)
Comment (nooodle-soup) [reply src/sql/magic.py]: Made it into a function
[function link](https://github.com/ploomber/jupysql/pull/878/files#diff-527caf90dd08297f5253e826bd98fea0b6f4fbd44fe9f20472976f77868dac06R689-R706)
Added tests
[test link](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2090-R2114)

File: src/tests/test_magic.py (hunk 17 above)
Comment (nooodle-soup) [reply src/tests/test_magic.py]: [Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2077-R2084)

**Comment by nooodle-soup** [comment]
@edublancas I have made the changes but some tests are failing and I cannot figure out why.
Can you please take a look and let me know where the issue is?

**Comment by edublancas** [comment]
I looked at the test:

```
        result = ip_with_dynamic_db.run_cell(
            f"""%%sql
    SELECT * FROM {__TABLE_NAME__};
    """
        ).result
    
>       assert len(result) == 3
E       TypeError: object of type 'NoneType' has no len()

```

`result` is `None`, and `len(None)` breaks. I'd first try investigating if your code introduced the issue or if it's an issue with an external dependency (you can try reproducing it locally or open a dummy PR to trigger the CI and see if tests pass or break with the same error).

**Comment by nooodle-soup** [comment]
@edublancas I've fixed the issues
Please have a look and let me know if everything is okay
Thanks!

File: src/sql/magic.py [hunk 18]
```diff
@@ -640,6 +674,38 @@ def _persist_dataframe(
         )
 
 
+def get_query_type(command: str):
+    """
+    Returns the query type of the original sql command
+    """
+    query_type = (
+        sqlparse.parse(command)[0].get_type() if sqlparse.parse(command) else None
```
Comment (edublancas) [inline src/sql/magic.py:682]: this is inefficient because it'll run sqlparse.parse twice, better to execute it only once

File: src/sql/magic.py [hunk 19]
```diff
@@ -640,6 +674,38 @@ def _persist_dataframe(
...
+    )
+    if query_type == "UNKNOWN":
+        return None
+    return query_type
+
+
+def get_dependency_in_CTE(sql_original, dependencies):
+    """
+    The following bit of code checks for dependencies within the parentheses
```
Comment (edublancas) [inline src/sql/magic.py:691]: this docstring is pretty long, we can remove things like "the following bit of code checks"

File: src/sql/magic.py [hunk 20]
```diff
@@ -640,6 +674,38 @@ def _persist_dataframe(
...
+    """
+    query_type = (
+        sqlparse.parse(command)[0].get_type() if sqlparse.parse(command) else None
+    )
+    if query_type == "UNKNOWN":
+        return None
+    return query_type
+
+
+def get_dependency_in_CTE(sql_original, dependencies):
```
Comment (edublancas) [inline src/sql/magic.py:689]: do not use capital letters in function names

File: src/sql/magic.py [hunk 21]
```diff
@@ -640,6 +674,38 @@ def _persist_dataframe(
...
+    """
+    The following bit of code checks for dependencies within the parentheses
+    We do not care about the name outside the parentheses being the same as
+    another table / snippet here and only check if there are dependencies
+    in the parentheses.
+    This is for generating more relevant warnings.
+    """
+
```
Comment (edublancas) [inline src/sql/magic.py:697]: I'm confused about what this function does. based on the code, this is executed when the `is_cte` flag is True, hence, it executes when the user's query begins with `WITH`, and it'll parse anything that's within parenthesis

e.g.,

```sql
with something (
 /* grab everything from here */
)
```

is this right?

if so, I'm confused because of the unit tests has a `CREATE TABLE AS ( ...)` as input, and it doesn't seem like this function will ever be executed with a `CREATE TABLE AS (...)` statement

Also, the unit tests only contain check when the CTE contains a single statement, but this implementation will yield weird results when there is more than one:


```
In [8]: sql_original = """
   ...:
   ...: with something as(
   ...: a
   ...: ), another as(
   ...: b)
   ...: """

In [9]:

In [9]: parentheses_content = re.findall(r"\((.*)\)", sql_original.replace("\n", " "))

In [10]: parentheses_content
Out[10]: [' a ), another as( b']
```

do we want `parenthesis_content` to be equal to `[' a ), another as( b']`? seems like an odd result.

I think at this point, it might be better to just delete this code and show a generic warning, because from what I see, the only purpose of this is to show better warnings.

**Review by edublancas — CHANGES_REQUESTED** [review]

File: setup.py [hunk 22]
```diff
@@ -18,7 +18,7 @@
 install_requires = [
     "prettytable",
     # IPython dropped support for Python 3.8
-    "ipython<=8.12.0; python_version <= '3.8'",
+    "ipython; python_version <= '3.8'",
```
[the remark this answers is not in the document]
Comment (nooodle-soup) [reply setup.py:21]: This change stopped the breaking issues for some tests
I'll investigate more and give an update.

File: src/sql/magic.py (hunk 21 above)
Comment (nooodle-soup) [reply src/sql/magic.py:697]: @edublancas I did write this logic for better warning only. I can improve it if you think we need detailed warnings, otherwise I can show a generic one.

**Comment by edublancas** [comment]
@nooodle-soup what's the status here?

**Comment by nooodle-soup** [comment]
@edublancas I'm waiting for a review
I've fixed all the issues that we discussed. I removed the get_dependencies_in_CTE and instead provided a more generic warning message.

File: setup.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (nooodle-soup) [reply setup.py]: The errors disappeared with the latest update of IPython

File: setup.py [hunk 23]
```diff
@@ -37,7 +37,7 @@
     "polars==0.17.2",  # 04/18/23 this breaks our CI
     "pyarrow",
     "invoke",
-    "pkgmt",
+    "pkgmt==0.7.3",
```
[the remark this answers is not in the document]
Comment (nooodle-soup) [reply setup.py]: I've removed it as the latest IPython update fixed it

File: src/sql/magic.py (hunk 18 above)
Comment (nooodle-soup) [reply src/sql/magic.py]: [Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-527caf90dd08297f5253e826bd98fea0b6f4fbd44fe9f20472976f77868dac06R677-R678)

File: src/sql/magic.py (hunk 19 above)
Comment (nooodle-soup) [reply src/sql/magic.py]: I've removed the function entirely and provided a better, more generic warning instead

File: src/sql/magic.py (hunk 20 above)
Comment (nooodle-soup) [reply src/sql/magic.py]: This function is no longer in use as I've removed it and provided a more generic warning.

File: src/sql/magic.py (hunk 21 above)
Comment (nooodle-soup) [reply src/sql/magic.py]: @edublancas I've removed the logic and provided a more generic warning instead

**Comment by edublancas** [comment]
@nooodle-soup remember to follow the guidelines and respond each PR observation with a link: https://ploomber-contributing.readthedocs.io/en/latest/contributing/responding-pr-review.html

also, you need to explicitly request a review, otherwise I don't know if the PR is ready for another review or not. please request the review either by using the github button or by adding a comment and tagging me

**Comment by nooodle-soup** [comment]
> @nooodle-soup remember to follow the guidelines and respond each PR observation with a link: https://ploomber-contributing.readthedocs.io/en/latest/contributing/responding-pr-review.html

@edublancas I thought I added links to almost all the responses where code was modified. Should I be adding links to the removed code as well? I assumed the code no longer existed so I should not be adding a link to that change 😓 . I'll add links there as well if you say so.

I'll ask for reviews by tagging from now on! 😄

**Comment by edublancas** [comment]
if you're removing code. you can just add a link to the commit that deleted it. I do that (one commit per PR observation and then respond to the PR reviewer with links to each commit)

**Comment by nooodle-soup** [comment]
> if you're removing code. you can just add a link to the commit that deleted it. I do that (one commit per PR observation and then respond to the PR reviewer with links to each commit)

Okay. I'll do the same from now on! 👍🏽

File: setup.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (nooodle-soup) [reply setup.py]: @edublancas  I don't have the link for the commit where I fixed this as I had to reset the git repo in my fork and pulled from master again to create a new branch that did not have this change.

File: src/sql/magic.py (hunk 19 above)
Comment (nooodle-soup) [reply src/sql/magic.py]: [Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-527caf90dd08297f5253e826bd98fea0b6f4fbd44fe9f20472976f77868dac06R449-R451)

File: src/sql/magic.py (hunk 20 above)
Comment (nooodle-soup) [reply src/sql/magic.py]: [Fixed](https://github.com/ploomber/jupysql/pull/878/commits/a204e262aa93b6bf06494dd2e06c1f27b25e0878)

File: src/sql/magic.py (hunk 21 above)
Comment (nooodle-soup) [reply src/sql/magic.py]: [Fixed](https://github.com/ploomber/jupysql/pull/878/commits/a204e262aa93b6bf06494dd2e06c1f27b25e0878)

**Comment by nooodle-soup** [comment]
@edublancas  I've added links to all the changes I could. Please have a look when you can and let me know if I can make my review better! Thanks!

File: src/tests/test_magic.py [hunk 24]
```diff
@@ -1987,3 +1987,102 @@ def test_accessing_previously_nonexisting_file(ip_empty, tmp_empty, capsys):
...
+SELECT * FROM my_penguins",
+            "SELECT",
+        ),
+        (
+            " WITH my_penguins ( \
+    SELECT * FROM penguins.csv \
+) \
+SELECT * FROM my_penguins",
+            None,
+        ),
```
Comment (edublancas) [inline src/tests/test_magic.py:2084]: this example is pretty hard to read, took me sometime to realize it's missing the `AS`. if you're testing invalid SQL, add some obviously invalid SQL code so the reader doesn't have to guess what you're testing

File: setup.py [hunk 25]
```diff
@@ -19,7 +19,7 @@
     "prettytable",
     # IPython dropped support for Python 3.8
     "ipython<=8.12.0; python_version <= '3.8'",
-    "ipython<8.16.0",
+    "ipython",
```
Comment (edublancas) [inline setup.py:22]: remove this, we already have ipython as dependency (line 21), we don't need this one

File: src/tests/test_magic.py [hunk 26]
```diff
@@ -1987,3 +1987,102 @@ def test_accessing_previously_nonexisting_file(ip_empty, tmp_empty, capsys):
...
+    print(out)
+    assert expected_result in out
+
+
+@pytest.mark.parametrize(
+    "query, query_type",
+    [
+        (
+            "CREATE TABLE penguins AS ( \
+    WITH my_penguins AS ( \
```
Comment (edublancas) [inline src/tests/test_magic.py:2064]: do not use `\` for splitting strings as it makes the tests hard to read, better to define multiline strings with `"""`

**Review by edublancas — CHANGES_REQUESTED** [review]

**Comment by edublancas** [comment]
@neelasha23 please run some manual tests and let us know if everything is ok

File: src/tests/test_magic.py (hunk 26 above)
Comment (nooodle-soup) [reply src/tests/test_magic.py]: [Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2072-R2107)

File: setup.py (hunk 25 above)
Comment (nooodle-soup) [reply setup.py]: [Fixed](https://github.com/ploomber/jupysql/pull/878/commits/3cce632b8d4c8f162b3a09b87c6fade4db0d15e4)

File: src/tests/test_magic.py (hunk 24 above)
Comment (nooodle-soup) [reply src/tests/test_magic.py]: [Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2097-R2100)

**Comment by nooodle-soup** [comment]
@edublancas 
I've implemented the changes and the PR is ready for review
Thanks!

File: src/tests/test_magic.py (hunk 24 above)
Comment (edublancas) [reply src/tests/test_magic.py]: is this link right? it's taking me to all the changed files, not to a specific set of lines or file

File: setup.py (hunk 25 above)
Comment (edublancas) [reply setup.py]: the link seems broken

image

File: src/tests/test_magic.py (hunk 26 above)
Comment (edublancas) [reply src/tests/test_magic.py]: please change it to `"""` (triple quotes)

File: setup.py (hunk 25 above)
Comment (nooodle-soup) [reply setup.py]: [Fixed](https://github.com/ploomber/jupysql/pull/878/commits/d6002ac0e223e15ca368c27bf22b6bf70d821158)

Updated the link

File: src/tests/test_magic.py (hunk 26 above)
Comment (nooodle-soup) [reply src/tests/test_magic.py]: [Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2059-R2092)

File: src/tests/test_magic.py (hunk 24 above)
Comment (nooodle-soup) [reply src/tests/test_magic.py]: [Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2082-R2090)
Updated link @edublancas

File: setup.py [hunk 27]
```diff
@@ -19,7 +19,6 @@
     "prettytable",
     # IPython dropped support for Python 3.8
     "ipython<=8.12.0; python_version <= '3.8'",
-    "ipython",
```
Comment (neelasha23) [inline setup.py:22]: why is ipython removed?

Comment (nooodle-soup) [reply setup.py:22]: @edublancas suggested the removal
Is there any issue you faced in your manual tests?

**Comment by neelasha23** [comment]
1. This still seems to throw an error and it's not very clear what should be the user's next step:
Screenshot 2023-10-10 at 12 13 30 PM

Specifying `--with` will also throw error:

Screenshot 2023-10-10 at 12 21 09 PM

2. It displays warning that CTE generation is disabled but the query is displaying some results. How is this working?
Screenshot 2023-10-10 at 12 31 27 PM

3. Here also it's not clear what the user needs to do as the next step
Screenshot 2023-10-10 at 12 33 04 PM

**Comment by nooodle-soup** [comment]
@neelasha23

If you saved a snippet called `my_penguins`, and also used it in a CTE in some way like 
```
WITH my_penguins AS (
    SELECT * FROM penguins.csv
) SELECT * FROM my_penguins
```

We want the code to run here because it is valid SQL that does not depend on any external snippet. Even though the name is the same, the `my_penguins` snippet is not to be used here and instead `my_snippet` temporary table which refers to `SELECT * FROM penguins.csv` is to be used.

**Comment by neelasha23** [comment]
> @neelasha23
> 
> If you saved a snippet called `my_penguins`, and also used it in a CTE in some way like
> 
> ```
> WITH my_penguins AS (
>     SELECT * FROM penguins.csv
> ) SELECT * FROM my_penguins
> ```
> 
> We want the code to run here because it is valid SQL that does not depend on any external snippet. Even though the name is the same, the `my_penguins` snippet is not to be used here and instead `my_snippet` temporary table which refers to `SELECT * FROM penguins.csv` is to be used.

So maybe we should display warning something like:  Running query with temporary_table `my_snippet` instead of saved snippet `my_snippet` . wdyt @edublancas

**Comment by nooodle-soup** [comment]
@neelasha23 I've updated my previous point

**Review by edublancas — APPROVED** [review]
